File: libembosser-core-drivers/src/main/java/org/brailleblaster/libembosser/drivers/utils/BaseTextEmbosser.java
package org.brailleblaster.libembosser.drivers.utils;

import java.io.IOException;
import java.io.InputStream;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;
import java.util.Optional;
import java.util.function.Function;

import javax.print.Doc;
import javax.print.DocFlavor;
import javax.print.DocPrintJob;
import javax.print.PrintException;
import javax.print.PrintService;
import javax.print.SimpleDoc;
import javax.print.StreamPrintServiceFactory;
import javax.print.event.PrintJobEvent;
import javax.print.event.PrintJobListener;

import org.brailleblaster.libembosser.drivers.utils.DocumentParser.ParseException;
import org.brailleblaster.libembosser.drivers.utils.document.events.DocumentEvent;
import org.brailleblaster.libembosser.spi.EmbossException;
import org.brailleblaster.libembosser.spi.Embosser;
import org.brailleblaster.libembosser.spi.EmbossingAttributeSet;
import org.brailleblaster.libembosser.spi.Rectangle;
import org.brailleblaster.libembosser.utils.EmbossToStreamPrintServiceFactory;
import org.jetbrains.annotations.NotNull;
import org.w3c.dom.Document;

import com.google.common.io.ByteSource;

public abstract class BaseTextEmbosser implements Embosser {
	private final StreamPrintServiceFactory streamPrintServiceFactory = new EmbossToStreamPrintServiceFactory();
	private final String id;
	private final String manufacturer;
	private final String model;
	private final Rectangle maximumPaper;
	private final Rectangle minimumPaper;
	// Not really sure whether this is the best way to monitor for print job finishing.
	// Doing it as this for now as this is how it worked previously in BrailleBlaster and so want too not break things.
	private boolean jobFinished;
	public static final byte ESC = 0x1B;
	public BaseTextEmbosser(String id, String manufacturer, String model, Rectangle maxPaper, Rectangle minPaper) {
		this.id = id;
		this.manufacturer = manufacturer;
		this.model = model;
		this.maximumPaper = maxPaper;
		this.minimumPaper = minPaper;
	}
	@Override
	public String getId() {
		return id;
	}
	@Override
	public String getManufacturer() {
		return manufacturer;
	}

	@Override
	public String getModel() {
		return model;
	}
	
	@Override
	public Rectangle getMaximumPaper() {
		return maximumPaper;
	}
	
	@Override
	public Rectangle getMinimumPaper() {
		return minimumPaper;
	}
	abstract protected @NotNull Function<Iterator<DocumentEvent>, ByteSource> createHandler(@NotNull EmbossingAttributeSet attributes);
	@Override
	public void embossPef(@NotNull PrintService embosserDevice, @NotNull Document pef, EmbossingAttributeSet attributes)
			throws EmbossException {
		DocumentParser parser = new DocumentParser();
		emboss(embosserDevice, pef, parser::parsePef, createHandler(attributes));
	}

	@Override
	public void embossBrf(@NotNull PrintService embosserDevice, @NotNull InputStream brf, EmbossingAttributeSet attributes)
			throws EmbossException {
		DocumentParser parser = new DocumentParser();
		emboss(embosserDevice, brf, parser::parseBrf, createHandler(attributes));
	}
	protected <T> boolean emboss(PrintService embosserDevice, T input, ThrowingBiConsumer<T, DocumentHandler, ParseException> parseMethod, Function<Iterator<DocumentEvent>, ByteSource> handler) throws EmbossException {
		List<DocumentEvent> events = new LinkedList<>();
		try {
			parseMethod.accept(input, events::add);
		} catch (ParseException e) {
			throw new EmbossException(e);
		}
		InputStream embosserStream;
		try {
			embosserStream = handler.apply(events.iterator()).openStream();
		} catch (IOException e) {
			throw new EmbossException(e);
		}
		return embossStream(embosserDevice, embosserStream);
	}
	/**
	 * A helper method for sending an InputStream to a printer device.
	 * 
	 * @param embosserDevice The printer device representing the embosser.
	 * @param is The InputStream to send to the embosser.
	 * @return True if the print job is successful false if there is a problem.
	 * @throws EmbossException Thrown if there is a problem embossing.
	 */
	protected boolean embossStream(PrintService embosserDevice, InputStream is) throws EmbossException {
		Doc doc = new SimpleDoc(is, DocFlavor.INPUT_STREAM.AUTOSENSE, null);
		DocPrintJob dpj = embosserDevice.createPrintJob();
		try {
			dpj.print(doc, null);
		} catch (PrintException e) {
			throw new EmbossException("Problem sending document to printer device", e);
		}
		dpj.addPrintJobListener(new PrintJobListener() {
			
			@Override
			public void printJobRequiresAttention(PrintJobEvent pje) {
				jobFinished = false;
			}
			
			@Override
			public void printJobNoMoreEvents(PrintJobEvent pje) {
				jobFinished = true;
			}
			
			@Override
			public void printJobFailed(PrintJobEvent pje) {
				jobFinished = true;
			}
			
			@Override
			public void printJobCompleted(PrintJobEvent pje) {
				jobFinished = true;
			}
			
			@Override
			public void printJobCanceled(PrintJobEvent pje) {
				jobFinished = true;
			}
			
			@Override
			public void printDataTransferCompleted(PrintJobEvent pje) {
				jobFinished = true;
			}
		});
		return jobFinished;
	}
	
	@Override
	public Optional<StreamPrintServiceFactory> getStreamPrintServiceFactory() {
		return Optional.of(streamPrintServiceFactory);
	}
}

File: libembosser-core-drivers/src/main/kotlin/org/brailleblaster/libembosser/drivers/utils/BaseTextEmbosser.kt
package org.brailleblaster.libembosser.drivers.utils

import com.google.common.io.ByteSource
import org.brailleblaster.libembosser.drivers.utils.document.events.DocumentEvent
import org.brailleblaster.libembosser.spi.EmbossException
import org.brailleblaster.libembosser.spi.Embosser
import org.brailleblaster.libembosser.spi.EmbossingAttributeSet
import org.brailleblaster.libembosser.spi.Rectangle
import org.brailleblaster.libembosser.utils.EmbossToStreamPrintServiceFactory
import org.w3c.dom.Document
import java.io.IOException
import java.io.InputStream
import java.util.*
import java.util.function.Function
import javax.print.*
import javax.print.event.PrintJobEvent
import javax.print.event.PrintJobListener

abstract class BaseTextEmbosser(private val id: String, private val manufacturer: String, private val model: String, private val maximumPaper: Rectangle, private val minimumPaper: Rectangle) : Embosser {
    private val streamPrintServiceFactory: StreamPrintServiceFactory = EmbossToStreamPrintServiceFactory()

    // Not really sure whether this is the best way to monitor for print job finishing.
    // Doing it as this for now as this is how it worked previously in BrailleBlaster and so want too not break things.
    private var jobFinished = false
    override fun getId(): String {
        return id
    }

    override fun getManufacturer(): String {
        return manufacturer
    }

    override fun getModel(): String {
        return model
    }

    override fun getMaximumPaper(): Rectangle {
        return maximumPaper
    }

    override fun getMinimumPaper(): Rectangle {
        return minimumPaper
    }

    protected abstract fun createHandler(attributes: EmbossingAttributeSet): Function<Iterator<DocumentEvent?>?, ByteSource>
    @Throws(EmbossException::class)
    override fun embossPef(embosserDevice: PrintService, pef: Document, attributes: EmbossingAttributeSet) {
        val parser = DocumentParser()
        emboss(embosserDevice, pef, { inputDoc: Document, handler: DocumentHandler -> parser.parsePef(inputDoc, handler) }, createHandler(attributes))
    }

    @Throws(EmbossException::class)
    override fun embossBrf(embosserDevice: PrintService, brf: InputStream, attributes: EmbossingAttributeSet) {
        val parser = DocumentParser()
        emboss(embosserDevice, brf, { input: InputStream, handler: DocumentHandler -> parser.parseBrf(input, handler) }, createHandler(attributes))
    }

    @Throws(EmbossException::class)
    protected fun <T> emboss(embosserDevice: PrintService, input: T, parseMethod: ThrowingBiConsumer<T, DocumentHandler, DocumentParser.ParseException?>, handler: Function<Iterator<DocumentEvent?>?, ByteSource>): Boolean {
        val events: MutableList<DocumentEvent> = LinkedList()
        try {
            parseMethod.accept(input, DocumentHandler { e: DocumentEvent -> events.add(e) })
        } catch (e: DocumentParser.ParseException) {
            throw EmbossException(e)
        }
        val embosserStream: InputStream = try {
            handler.apply(events.iterator()).openStream()
        } catch (e: IOException) {
            throw EmbossException(e)
        }
        return embossStream(embosserDevice, embosserStream)
    }

    /**
     * A helper method for sending an InputStream to a printer device.
     *
     * @param embosserDevice The printer device representing the embosser.
     * @param is The InputStream to send to the embosser.
     * @return True if the print job is successful false if there is a problem.
     * @throws EmbossException Thrown if there is a problem embossing.
     */
    @Throws(EmbossException::class)
    protected fun embossStream(embosserDevice: PrintService, `is`: InputStream): Boolean {
        val doc: Doc = SimpleDoc(`is`, DocFlavor.INPUT_STREAM.AUTOSENSE, null)
        val dpj = embosserDevice.createPrintJob()
        try {
            dpj.print(doc, null)
        } catch (e: PrintException) {
            throw EmbossException("Problem sending document to printer device", e)
        }
        dpj.addPrintJobListener(object : PrintJobListener {
            override fun printJobRequiresAttention(pje: PrintJobEvent) {
                jobFinished = false
            }

            override fun printJobNoMoreEvents(pje: PrintJobEvent) {
                jobFinished = true
            }

            override fun printJobFailed(pje: PrintJobEvent) {
                jobFinished = true
            }

            override fun printJobCompleted(pje: PrintJobEvent) {
                jobFinished = true
            }

            override fun printJobCanceled(pje: PrintJobEvent) {
                jobFinished = true
            }

            override fun printDataTransferCompleted(pje: PrintJobEvent) {
                jobFinished = true
            }
        })
        return jobFinished
    }

    override fun getStreamPrintServiceFactory(): Optional<StreamPrintServiceFactory> {
        return Optional.of(streamPrintServiceFactory)
    }

    companion object {
        const val ESC: Byte = 0x1B
    }
}